dramatiq/composition.py

```diff
@@ -224,7 +242,51 @@ def run(self, *, delay=None):
           delay(int): The minimum amount of time, in milliseconds,
             each message in the group should be delayed by.
         """
-        for child in self.children:
+        if self.completion_callbacks:
+            from .middleware.group_callbacks import GROUP_CALLBACK_BARRIER_TTL, GroupCallbacks
+
+            rate_limiter_backend = None
+            for middleware in self.broker.middleware:
+                if isinstance(middleware, GroupCallbacks):
+                    rate_limiter_backend = middleware.rate_limiter_backend
+                    break
+            else:
+                raise RuntimeError(
+                    "GroupCallbacks middleware not found! Did you forget "
+                    "to set it up? It is required if you want to use "
+                    "group callbacks."
+                )
+
+            # Generate a new completion uuid on every run so that if a
+            # group is re-run, the barriers are all separate.
+            # Re-using a barrier's name is an unsafe operation.
+            completion_uuid = str(uuid4())
+            completion_barrier = Barrier(rate_limiter_backend, completion_uuid, ttl=GROUP_CALLBACK_BARRIER_TTL)
+            completion_barrier.create(len(self.children))
+
+            children = []
+            for child in self.children:
+                if isinstance(child, group):
+                    raise NotImplementedError
+
+                elif isinstance(child, pipeline):
+                    pipeline_children = child.messages[:]
+                    pipeline_children[-1] = pipeline_children[-1].copy(options={
+                        "group_completion_uuid": completion_uuid,
+                        "group_completion_callbacks": self.completion_callbacks,
+                    })
+
+                    children.append(pipeline(pipeline_children, broker=child.broker))
+
+                else:
+                    children.append(child.copy(options={
+                        "group_completion_uuid": completion_uuid,
```

Comment:
Groups of groups, unlike pipelines of pipelines, cannot be reduced, because there will a functional distinction when results are eventually considered, which I'm sure is an eventual goal of this feature, though I would try to leave it to a later iteration.

To that end, you'll need to make this property take many values, and because of that I'd recommend changing the name to `group_completion_callbacks`, and having it be a list instead of a string. Even if you decide to punt on the group-of-groups feature for now and just leave it `NotImplemented`, I'd encourage changing the name so that we don't need to write compatiblity code that will need to handle multiple names in the future.